compiler/rustc_hir_analysis/src/astconv/mod.rs

```diff
@@ -3680,6 +3680,19 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                         Applicability::MachineApplicable,
                     );
                 }
+                // note if the user wants to use associated items of a trait directly
+                if in_path && res.len() == 0 {
+                    if let Ok(snippet) = tcx.sess.source_map().span_to_snippet(self_ty.span) {
+                        diag.note(format!(
+                            "desired associated item not found in trait `{}` if you want to use such one",
```

Comment:
I'm confused by this wording. It also seems redundant given the note that already exists: `no function or associated item named `has` found for trait object `dyn HasNot` in the current scope`

Reply:
This wording says that there is no direct associated item in this trait, and the `dyn HasNot` in the existing note causes confusion (even though `dyn HasNot` also contains associated item of `HasNot`)

Reply:
I'm not sure if I understand what you mean by this, and I still think that this note is not adding any new information over the existing note.

Reply:
How about this: "no desired associated item found in this trait, if you do not want to use a bare trait object here".

Reply:
The `trait object `dyn HasNot`` in the existing note will confuse users.